Add E2E tests for new tag list

jonatansberg opened the pull request:
These PR adds some basic tests for the new tags screen to the new E2E playwright setup. 

There are still some mocking in here since the test are not yet isolated and run against the development env locally. I'd love to get rid of that once we have the testing env sorted!

jonatansberg committed:
Set up basic E2E tests for tag list

Refs https://linear.app/ghost/issue/PROD-2508/set-up-tests-within-new-e2e-structure

Uses the new E2E setup. Needs some mocking still since tests are not run in 
isolation and locyllt they run against your dev environment. Once isolation is 
in place we can remove most (if not all) mocks.

jonatansberg committed:
Add tests for infinite scrolling

Refs https://linear.app/ghost/issue/PROD-2508/set-up-tests-within-new-e2e-structure

Uses mocking since test isolation and data creation factories are not yet ready.

jonatansberg commented:
Seems like the feature flag isnt enabled when you run the tests in CI... Thats why thet are failing.

jonatansberg committed:
Ensure feature flag is enabled during testing

Refs https://linear.app/ghost/issue/PROD-2508/set-up-tests-within-new-e2e-structure

Using Playwright to amend the settings response lets us ensure the necessary 
feature flag is enabled in the context of this test without affecting the 
environment in which we are running (e.g. local dev, ci, etc.)

jonatansberg commented on e2e/helpers/pages/admin/tags/TagsPage.ts:
```diff
@@ -0,0 +1,79 @@
+
+    // XXX: Remove once we have proper test isolation and don't need mocking
+    async mockTagsResponse(handler: (request: Request) => Promise<Partial<PaginatedResponse>>) {
+        await this.page.route('/ghost/api/admin/tags/*', async (route, request) => {
+            const tags = await handler(request);
+            await route.fulfill({
+                body: JSON.stringify({
+                    meta: {
+                        ...tags.meta,
+                        pagination: {
+                            page: 1,
+                            limit: 100,
+                            pages: 1,
+                            total: tags.meta?.pagination?.total ?? tags.tags?.length ?? 0,
+                            ...tags.meta?.pagination
+                        }
+                    },
+                    tags: tags.tags ?? []
+                })
+            });
+        });
```
Works as is

jonatansberg commented on e2e/helpers/pages/admin/tags/TagsPage.ts:
```diff
@@ -0,0 +1,79 @@
+
+// XXX: Remove these types and the mockTagsResponse method once we have proper
+// test isolation and no longer need to mock responses.
+interface Tag {
+    id: string;
+    name: string;
+    slug: string;
+    url: string;
+    description: string;
+}
```
Wont do

9larsons reviewed (changes requested):
Sorry for the delay on reviewing this. All in all, I'm stoked that these work! We're very close to being able to hit the ground running with this package.

I don't think these tests are critical to merge as it's more of a spike. Given that, I'd like to:
1) Rebase to pull in test isolation changes
2) Lean more on locators and the Page object(s)

I feel less strongly about the need for these to happen before merging, but I do feel strongly that they should happen:
- (Create and) Utilize a TagFactory
- Talk w/ Igor about assertions; I suspect he'll want to revise here, but I think this is a great basis for discussion on approach for us all to align on

jonatansberg commented on e2e/helpers/pages/admin/tags/TagsPage.ts:
```diff
@@ -0,0 +1,78 @@
+        this.pageContent = page.getByTestId('tags-page');
+        this.tagList = page.getByTestId('tags-list');
+
+        // XXX: This is the only way I have found to reliably selecting the table rows 
```
I had that at one point, but I ran in to some kind of issue... Coming back to this, I can't see why that wouldn't work tho 🤔

jonatansberg commented on e2e/tests/admin/tags/list.test.ts:
```diff
@@ -0,0 +1,227 @@
+        // Default to public tags
+        await expect(tagsPage.activeTab).toHaveText('Public tags');
+
+        await expect(tagsPage.tagList).toBeVisible();
```
That seems reasonable. I'll drop that assertion.

ibalosh reviewed (changes requested):
Hey fellas,

I went through the PR to check bit more things. I think it needs cleanup, to get to our goal of nice tests that we and AI will like too, when it starts generating new ones.

Shared some of my thoughts in there.

jonatansberg committed:
Move all locators to page object

Refs https://linear.app/ghost/issue/PROD-2508/set-up-tests-within-new-e2e-structure

This makes the tags tests more consistent with the other E2E tests at the cost 
of some extra indirection.

jonatansberg committed:
Move request overrides to separate helper files

Refs https://linear.app/ghost/issue/PROD-2508/set-up-tests-within-new-e2e-structure

This makes the implementation cleaner by co-locating the request overrides while
keeping the tags page object focused on page manipulation and interaction.

jonatansberg committed:
Reword XXX comments as TODOs

Refs https://linear.app/ghost/issue/PROD-2508/set-up-tests-within-new-e2e-structure

We don't use XXX comments in the rest of the code base. Sticking with TODOs 
keeps things more consistent.

jonatansberg committed:
Make overrideFeatureFlags a shared helper

Refs https://linear.app/ghost/issue/PROD-2508/set-up-tests-within-new-e2e-structure

The overrideFeatureFlags helper has utility outside of the tags tests. Moving 
it to the shared folder makes it easier to reuse.

ibalosh commented on e2e/helpers/pages/admin/tags/TagsPage.ts:
```diff
@@ -0,0 +1,44 @@
+        this.newTagButton = page.getByRole('link', {name: 'New tag'});
+
+        this.emptyStateTitle = this.pageContent.getByRole('heading', {name: 'Start organizing your content'});
+        this.emptyStateAction = this.pageContent.getByRole('link', {name: 'Create a new tag'});
```
since we use button in other places we should here too.

```suggestion
this.createNewTagButton = this.pageContent.getByRole('link', {name: 'Create a new tag'});
```

jonatansberg replied:
We're already using the `newTagButton` in the test that I think you are referring to? This is a different button that only shows in the empty state. We could remove it, though I don't think it adds much value.

jonatansberg replied:
Ah, okay, I see. 

Even though perhaps it's more consistent I also find it a bit confusing. I wouldn't know how to tell them apart without looking at both the locator definition and the UI. The `emptyStateAction` name more clearly conveys the semantic meaning of that button to me and it also avoids that conflict. 

I guess you could call it the `emptyStateActionButton`, but that just gets wordy 😅

ibalosh reviewed (approved):
I added couple of comments here and there, but overall I think it's good. We can cleanup further as we add more and introduce tags factory.

One thing though, before you merge it, make sure it passes - I did not get the tests to pass on my machine.
Ready to approve, but did not yet, just to make sure it passes first, once it does, ready to approve @jonatansberg

jonatansberg commented on e2e/tests/admin/tags/list.test.ts:
```diff
@@ -0,0 +1,227 @@
+
+        await tagsPage.newTagButton.click();
+
+        await expect(page.url()).toContain('/ghost/#/tags/new');
```
Having some kind of assertion that the URL changes is actually relevant in this case, since we're navigating across the React-Ember border at this point. Thats why we want to make sure that we're actually navigated away from the list page.

9larsons reviewed (approved):
As noted above, we're going to end up changing a lot of these tests and I didn't mean to hold up the posts app changes in here. We'll come back around once the Tags factory and Settings factory PRs are merged.

In general not a fan of how the tests read now, but it's not your fault. We need a few more pieces in place.
